### ISSUE TITLE
a Flag to deactivate sds fallback if a socket is not discovered. 

### ISSUE THREAD

**Issue by asayah** [istio/istio#45534]
Currently, Istio can pick up a socket when starting the sidecar for sds, To integrate with SPIRE for example, but in case of an issue it will fallback to the default sds, this can constitute a security risk, or resulting in different sidecars using different sds within the same mesh. 

**Describe the feature request**

we need a flag to deactivate the fallback to the default sds, (false by default) 

**Describe alternatives you've considered**
Script terminating a sidecar if a cert is not issued by SPIRE. 

**Affected product area (please put an X in all that apply)**

[ ] Ambient
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[X] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Additional context**
https://github.com/istio/istio/pull/37947

**Comment by howardjohn** [comment]
What types of issues are we expecting here? would an init container to check the socket exists make sense?

**Comment by howardjohn** [comment]
If the flag is set and socket is not what should we do? exit, wait,...?

**Comment by asayah** [comment]
I was thinking since it is critical, the pod should be just recycled when the socket is not found, and we can leverage the kube backoff mechanism to restart the pods (giving more time for the socket to be ready sometimes) 

tldr: exit

**Comment by asayah** [comment]
> 

exit

**Comment by asayah** [comment]
hoping to submit a quick PR this/next week for this.

**Comment by howardjohn** [comment]
The central point of the design was to not require a flag if I remember correctly. Were there just so bad assumptions made there that the socket would always be ready before the pod?

**Comment by asayah** [comment]
It is hard to reproduce this issue using a CSI driver, maybe the assumption was based on this, though there is still a lot of legacy implementation mounting the socket manually

**Comment by bleggett** [comment]
> It is hard to reproduce this issue using a CSI driver, maybe the assumption was based on this, though there is still a lot of legacy implementations of SPIRE mounting the socket manually

Considering that manual socket mounts are not secure and upstream SPIRE really only supports the CSI driver now, do we want to add config flags that are mooted if you're using the CSI driver (which everyone should be?)

**Comment by asayah** [comment]
> > It is hard to reproduce this issue using a CSI driver, maybe the assumption was based on this, though there is still a lot of legacy implementations of SPIRE mounting the socket manually
> 
> Considering that manual socket mounts are not secure and upstream SPIRE really only supports the CSI driver now, do we want to add config flags that are mooted if you're using the CSI driver (which everyone should be?)

I'm with you, if we are expecting sds we need to through an error without needing a flag, the problem is that today there is no way to say this sds socket is required, it is just checking if it exists (mounted) it will use it instead of starting istio sds. 

so we either need to define a way (through proxy configuration or mesh, to say use custom sds vs istio) or keep the current behavior and use a flag

**Comment by bleggett** [comment]
> > > It is hard to reproduce this issue using a CSI driver, maybe the assumption was based on this, though there is still a lot of legacy implementations of SPIRE mounting the socket manually
> > 
> > 
> > Considering that manual socket mounts are not secure and upstream SPIRE really only supports the CSI driver now, do we want to add config flags that are mooted if you're using the CSI driver (which everyone should be?)
> 
> I'm with you, if we are expecting sds we need to through an error without needing a flag, the problem is that today there is no way to say this sds socket is required, it is just checking if it exists (mounted) it will use it instead of starting istio sds.
> 
> so we either need to define a way (through proxy configuration or mesh, to say use custom sds vs istio) or keep the current behavior and use a flag

Ah I see.

Yep - ideally I'd rather just have the alt SDS path itself set via Helm/pod env, and if it's set/non-empty we bail if we can't find it.

Barring that, reworking the flag your PR to simply _be_ `UseExternalWorkloadSDS`  instead of `ExitIfSDSSocketNotFound` is good enough.

**Comment by asayah** [comment]
>Barring that, reworking the flag in your PR to simply be UseExternalWorkloadSDS instead of ExitIfSDSSocketNotFound is good enough.
Make sense, will rename the flag.
